### ISSUE TITLE
File reader and writer for .mps and .lp files (`scipy.optimize.linprog`)

### ISSUE THREAD

**Issue by jhelgert** [scipy/scipy#14593]
The [`linprog`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.linprog.html) function interfaces multiple LP solvers, including the quite promising HIGHS solver. Obviously, it requires the user to provide the LP in matrix form. However, in practice, LPs are distributed by standardised [.mps](https://www.gurobi.com/documentation/9.1/refman/mps_format.html) or [.lp](https://www.gurobi.com/documentation/9.1/refman/lp_format.html) files to enable users to solve the LP across different used languages, modelling libraries and solvers, see for example [Mittelmann's Benchmark library](http://plato.asu.edu/bench.html).

IMHO, it would be nice to have `mpsread(filename: str)` and `lpread(filename: str)` methods similar to MATLAB's [mpsread](https://de.mathworks.com/help/optim/ug/mpsread.html) to read a LP from a file which both return the matrix data of the optimization problem. Since it is already discussed to interface HiGHs' QP solver (#13480) and MILP solver (#14455), the same argument also applies for QPs and MILPs (both use the same file format).

I don't think it's necessary to write a file reader from scratch. From my point of view, the easiest way might be to write a Cython wrapper of [HiGHS' FileReader](https://github.com/scipy/scipy/tree/master/scipy/optimize/_highs/src/io) classes. Provided that there's interest and no one else is already working on something similar, I'd be happy to work on a PR. :)

**Comment by mdhaber** [comment]
We actually have been using two - the HiGHS one, as you mentioned, and one I wrote in Python before that was available. We just haven't made public interfaces for either, and this is the first request to do that, i think. @mckib2 what do you think?

**Comment by mckib2** [comment]
Yes, we used to build a private MPS file writer extension with HiGHS, but because it was unused it got removed in gh-13626.  Should be easy to resurrect and add in alongside the reader and LP file type functions as Cython wrappers like you suggest.  Would love to see a PR for this, happy to help review

**Comment by jhelgert** [comment]
> Yes, we used to build a private MPS file writer extension with HiGHS, but because it was unused it got removed in [gh-13626](https://github.com/scipy/scipy/pull/13626). Should be easy to resurrect and add in alongside the reader and LP file type functions as Cython wrappers like you suggest. Would love to see a PR for this, happy to help review

Nice, I'll start working on a PR based on your `_mpswriter` at the weekend! :)

**Comment by pmlpm1986** [comment]
This feature would be much appreciated. Any news on this front?

**Comment by mckib2** [comment]
xref gh-15217: updates HiGHS to latest master and that will presumably include some interface changes.  I haven't heard much on this front, @jhelgert is this something you're still working on?

**Comment by jhelgert** [comment]
Oops, I had forgotten this one, sorry! Unfortunately, a few things got in the way in August. However, I think I can work on a PR (based on #15217) in the next few days.

**Comment by rgommers** [comment]
It would be useful if this went into `scipy.io` I think, and not into `scipy.optimize`. It'll be a little fiddly to not introduce a dependency of `io` on `optimize` I think, given that the Cython wrapper is in `optimize`. Maybe okay to introduce that dependency, and possibly refactor later.

Does that sound right @mckib2?

**Comment by mckib2** [comment]
> Does that sound right @mckib2?

Yes, I agree this should go in `scipy.io`.  Since the HiGHS' sources are in `scipy/_lib/highs`, it can be written to have no dependency on `optimize` by just pulling and compiling the correct sources straight from there
